lib/OpenQA/WebAPI/Plugin/ObsRsync/Controller.pm

```diff
@@ -0,0 +1,161 @@
+    my $files
+      = $full->list({dir => 1, hidden => 1})->map(sub { $_->basename })->grep(qr/\.run_.*/)->sort(sub { $b cmp $a })
+      ->to_array;
+    $self->render('ObsRsync_logs', folder => $folder, full => $full->to_string, subfolders => $files);
+}
+
+sub logfiles {
+    my $self      = shift;
+    my $folder    = $self->param('folder');
+    my $subfolder = $self->param('subfolder');
+    return if $self->_check_and_render_error($folder, $subfolder);
+
+    my $full  = Mojo::File->new($self->_home, $folder, $subfolder);
+    my $files = $full->list({dir => 1})->map(sub { $_->basename })->sort->to_array;
+    $self->render(
+        'ObsRsync_logfiles',
+        folder    => $folder,
+        subfolder => $subfolder,
+        full      => $full->to_string,
+        files     => $files
+    );
+}
+
+sub download_file {
+    my $self      = shift;
+    my $folder    = $self->param('folder');
+    my $subfolder = $self->param('subfolder');
+    my $filename  = $self->param('filename');
+    return if $self->_check_and_render_error($folder, $subfolder, $filename);
+
+    my $full = $self->_home . '/' . $folder;
+    $full = $full . '/' . $subfolder if $subfolder;
+    $self->reply->file("$full/$filename");
+}
+
+sub run {
+    my $self   = shift;
+    my $folder = $self->param('folder');
+    return if $self->_check_and_render_error($folder);
+
+    my $cmd    = "bash '" . $self->_home . "/rsync.sh' '$folder' 2>&1";
+    my $out    = `$cmd`;
+    my $rc     = $? >> 8;
+    my $status = $rc ? 500 : 201;
+    return $self->render(json => {output => $out, code => $rc}, status => $status);
+}
+
+sub _grep_and_stash_scalar {
+    my ($self, $files, $mask, $var) = @_;
+    my $r = "";
+    my @r = grep /$mask/, @$files;
+    if (@r) {
+        $r = $r[0];
+    }
+    $self->stash($var, $r);
+}
+
+sub _grep_and_stash_list {
+    my ($self, $files, $mask, $var) = @_;
+    my @r = grep /$mask/, @$files;
+    $self->stash($var, \@r);
+}
+
+sub _check_and_render_error {
+    my ($res, $code) = _check_error(@_);
+    shift->_render_error($res, $code) if $res;
+    return $res;
+}
+
+sub _check_error {
+    my $self      = shift;
+    my $project   = shift;
+    my $subfolder = shift;
+    my $filename  = shift;
+    return ("Home directory is not set", 405) unless $self->_home;
+    return ("Home directory not found",  405) unless -d $self->_home;
+    return "Project has invalid characters" if $project && CORE::index($project, '/') != -1;
+    return "Subfolder has invalid characters" if ($subfolder && CORE::index($subfolder, '/') != -1);
+    return "Filename has invalid characters"  if ($filename  && CORE::index($filename,  '/') != -1);
+
+    return 404 unless !$project || -d $self->_home . '/' . $project;
+
+    return 0;
+}
+
+sub _render_error {
+    my $self    = shift;
+    my $message = shift;
+    my $code    = shift;
+
+    return $self->render(status => $message) if (($message + 0) eq $message);
```

Comment:
This could probably be done cleaner with better `_render_error` arguments. Having the message sometimes be numeric is strange.